prov/sharp: Implement sharp_domain2() using sharp_coll API

ldorau opened the pull request:
Signed-off-by: Lukasz Dorau <lukasz.dorau@intel.com>

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grom72/libfabric/4)
<!-- Reviewable:end -->

grom72 committed:
prov/sharp: SHARP provider scaffolding

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/coll: Prepare coll_cq to be used by offload collective providers

coll_cq implementation can be reused by other collective providers.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: collective offload capabilities reported if offload provider is available

* FOR TEST PURPOSE ONLY - TEMPORARY SOLUTION *
To let all collective tests pass let's force rxm to provide only offload collective
domain capabilities. THIS IS ONLY FOR TEST PPURPOSE. In the future, tests for
offload provider shall be executed with OFI_OFFLOAD_PROV_ONLY flag set in
fi_query_collective() call.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Collective offload fabric setup integrated into rxm fabric initialization

It is rxm provider responsability to initialize collective offload provider's fabric.
Otherwise collective offload functionality will not be available

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Create eq for collective offload provider

Peer provider must create peer_eq for offload provider, to allow offload provider
reporting events to peer provider.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
fabtests/multinode: Skip tests for unsupported collective operations

Some collective providers do not support all collective operations.
In such a case unsupported operations will not affect the final test result.

fi_query_collective() is used to check operation support.

A new function, test_query(), call fi_query_collective() with parameters defined in
extended coll_test structure.
The test is silently skipped if fi_query_collective returns -FI_ENOSYS.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Set offload_coll_mask based on actual configuration

offload_coll_mask value is calculated based on the actual offload capabilities
confirmed by fi_query_collective().

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: Report only coll offload capabilities when OFI_OFFLOAD_PROV_ONLY is set

fi_query_collective() reports only collective offload provider capabilities if
OFI_OFFLOAD_PROV_ONLY flag is set. Otherwise the sum of both providers' capabilities
is reported.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/rxm: rxm_mc introduction

rxm_mc added to hide implementation details of multicast address of collective
operations.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/util: fi_peer_mc_context for fi_join_collective to pass peer's mc

Collective provider (i.e. SHARP) requires to execute collective
operation on the supporting collective provider while running join() operation.
Such operation is executed on peer_ep with peer_mc as address of the operation.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/util: util_coll uses peer_mc_context

util_coll:fi_join() called with the FI_PEER flag restores peer_mc_context.mc_fid and
uses it as the actual context of fi_join() operation. This includes also reporting
the join operation completion with mc_fid as event's context.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/sharp: Dummy implementation of fi_barrier() and fi_allreduce()

Both collective operation implemented ba colling peer collective operation and
transparently passing completion back to peer CQ

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/sharp SHARP coll driver access API scaffolding

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/sharp: Sharp provider TODO list

TODO.md files defines next development steps. The list is ordered by priority.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

ldorau committed:
prov/sharp: Add mocks for sharp_coll_init and sharp_coll_finalize

Add mocks for sharp_coll_init and sharp_coll_finalize.

Signed-off-by: Lukasz Dorau <lukasz.dorau@intel.com>

ldorau committed:
prov/sharp: sharp provider can be mocked

In order to mock SHARP functions run:
$ ./configure --enable-sharp=mocked
or
$ ./configure --enable-sharp=auto

Running:
$ ./configure --enable-sharp=yes
make use of SHARP libraries instead of mocks.

Signed-off-by: Lukasz Dorau <lukasz.dorau@intel.com>

grom72 committed:
prov/sharp: Dummy implementation of fi_join_collective()

fi_join_collective implementation with usage of coll_op's state machine

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
prov/coll: Prepare coll_eq to be used by offload collective providers

coll_eq implementation can be reused by other collective providers.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 reviewed (approved):
> Reviewed 3 of 3 files at r1, 1 of 2 files at r2, 1 of 4 files at r5, 31 of 37 files at r6, 6 of 6 files at r7.
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: all files reviewed (commit messages unreviewed), 2 unresolved discussions (waiting on @ldorau)

___
*[`prov/sharp/src/sharp_domain.c` line 273 at r7](https://reviewable.io/reviews/grom72/libfabric/4#-NKXEqFd8RkQAZ7Pb9cp:-NKXEqFd8RkQAZ7Pb9cq:undefined) ([raw file](https://github.com/grom72/libfabric/blob/8e01c400ccfe93f1693a26a947349704edaab5e3/prov/sharp/src/sharp_domain.c#L273)):*
> ```C
> /* ??? */	.ib_dev_list = NULL,		/**< IB device name, port list. (const char *) */
> 		.user_progress_num_polls = -1,	/**< Number of polls to do before calling user progress. (int) */
> /* ??? */	.coll_timeout = 0,		/**< Timeout (msec) for collective operation, -1 - infinite (int) */
> ```





_Suggestion:_
```C
-1
```
___
*[`prov/sharp/src/sharp_domain.c` line 285 at r7](https://reviewable.io/reviews/grom72/libfabric/4#-NKXF9PJ2WhU4rut7HYE:-NKXF9PJ2WhU4rut7HYF:undefined) ([raw file](https://github.com/grom72/libfabric/blob/8e01c400ccfe93f1693a26a947349704edaab5e3/prov/sharp/src/sharp_domain.c#L285)):*
> ```C
> /* ??? */	.oob_colls = oob_colls,			/**< @ref sharp_coll_out_of_band_colls "List of OOB collectives". */
> 		.world_local_rank = 0,			/**< relative rank of this process on this node within its job. */
> /* ??? */	.enable_thread_support = 0,		/**< enable multi threaded support. */
> ```





_Suggestion:_
```C
1
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: all files reviewed, 2 unresolved discussions (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: all files reviewed, 2 unresolved discussions (waiting on @grom72)

___
*[`prov/sharp/src/sharp_domain.c` line 273 at r7](https://reviewable.io/reviews/grom72/libfabric/4#-NKXEqFd8RkQAZ7Pb9cp:-NKXS0jH3bRFDVSCaaiO:b-896fix) ([raw file](https://github.com/grom72/libfabric/blob/8e01c400ccfe93f1693a26a947349704edaab5e3/prov/sharp/src/sharp_domain.c#L273)):*
> ```C
> /* ??? */	.ib_dev_list = NULL,		/**< IB device name, port list. (const char *) */
> 		.user_progress_num_polls = -1,	/**< Number of polls to do before calling user progress. (int) */
> /* ??? */	.coll_timeout = 0,		/**< Timeout (msec) for collective operation, -1 - infinite (int) */
> ```

Done.



_Code quote:_
```C
coll_timeout = 0,	
```
___
*[`prov/sharp/src/sharp_domain.c` line 285 at r7](https://reviewable.io/reviews/grom72/libfabric/4#-NKXF9PJ2WhU4rut7HYE:-NKXSMfs5Pcc8kzNLaza:b-896fix) ([raw file](https://github.com/grom72/libfabric/blob/8e01c400ccfe93f1693a26a947349704edaab5e3/prov/sharp/src/sharp_domain.c#L285)):*
> ```C
> /* ??? */	.oob_colls = oob_colls,			/**< @ref sharp_coll_out_of_band_colls "List of OOB collectives". */
> 		.world_local_rank = 0,			/**< relative rank of this process on this node within its job. */
> /* ??? */	.enable_thread_support = 0,		/**< enable multi threaded support. */
> ```

Done.



_Code quote:_
```C
able_thread_support = 0
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (changes requested):
> Reviewed 1 of 1 files at r9, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: all files reviewed (commit messages unreviewed), 1 unresolved discussion (waiting on @ldorau)

___
*[`prov/sharp/src/sharp_domain.c` line 261 at r10](https://reviewable.io/reviews/grom72/libfabric/4#-NKZfdREEe-ekUwx_B3E:-NKZfdREEe-ekUwx_B3F:biwvkw4) ([raw file](https://github.com/grom72/libfabric/blob/f207916609b81bac30a32e22e09c289957949d3d/prov/sharp/src/sharp_domain.c#L261)):*
> ```C
> 
> 	struct sharp_coll_config config = {
> 		.ib_dev_list = strdup(domain_attr.name),/**< IB device name, port list. (const char *) */
> ```

.ib_dev_list must be released after it is used



_Code quote:_
```C
.ib_dev_list = strdup(domain_attr.name)
```


<!-- Sent from Reviewable.io -->

grom72 reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: all files reviewed (commit messages unreviewed), 2 unresolved discussions (waiting on @ldorau)

___
*[`prov/sharp/src/sharp_domain.c` line 261 at r10](https://reviewable.io/reviews/grom72/libfabric/4#-NKZg1gI8wNCnkxzheo2:-NKZg1gI8wNCnkxzheo3:b-vxj5s8) ([raw file](https://github.com/grom72/libfabric/blob/f207916609b81bac30a32e22e09c289957949d3d/prov/sharp/src/sharp_domain.c#L261)):*
> ```C
> 
> 	struct sharp_coll_config config = {
> 		.ib_dev_list = strdup(domain_attr.name),/**< IB device name, port list. (const char *) */
> ```

``ib_dev_list`` shall be given as *"``<device_name>``*``:``*``<port_number>``"* i.e. ``"mlx5_0:1"``
*``port_number``* is defined via parameter ``FI_SHARP_IB_PORT``
See ``sharp_init_env()``



_Code quote:_
```C
.ib_dev_list
```


<!-- Sent from Reviewable.io -->

grom72 committed:
prov/rxm: memory leaks removed

fi_close() path of domain and eq fixed in several places to avoid
memory leaks.
Obvious memory leaks removed from rxm_peer_av_ep_addr(), too.

Signed-off-by: Tomasz Gromadzki <tomasz.gromadzki@intel.com>

grom72 committed:
Merge branch 'prov-rxm-mc' into devel

grom72 committed:
Merge branch 'coll_test_skip_if_not_supported' into devel

ldorau committed:
prov/sharp: Implement sharp_domain2() using sharp_coll API

Signed-off-by: Lukasz Dorau <lukasz.dorau@intel.com>

ldorau reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: 37 of 38 files reviewed, 2 unresolved discussions (waiting on @grom72)

___
*[`prov/sharp/src/sharp_domain.c` line 261 at r10](https://reviewable.io/reviews/grom72/libfabric/4#-NKZfdREEe-ekUwx_B3E:-NKlwDTdCSwU5vsYNrWK:b-896fix) ([raw file](https://github.com/grom72/libfabric/blob/105ae920622bde1e4cfa7df7f27b3f0319d3eb49/prov/sharp/src/sharp_domain.c#L261)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

.ib_dev_list must be released after it is used
</blockquote></details>

Done.
___
*[`prov/sharp/src/sharp_domain.c` line 261 at r10](https://reviewable.io/reviews/grom72/libfabric/4#-NKZg1gI8wNCnkxzheo2:-NKlwEd-5_g3J6TMcz2W:b-896fix) ([raw file](https://github.com/grom72/libfabric/blob/105ae920622bde1e4cfa7df7f27b3f0319d3eb49/prov/sharp/src/sharp_domain.c#L261)):*
<details><summary><i>Previously, grom72 (Tomasz Gromadzki) wrote…</i></summary><blockquote>

``ib_dev_list`` shall be given as *"``<device_name>``*``:``*``<port_number>``"* i.e. ``"mlx5_0:1"``
*``port_number``* is defined via parameter ``FI_SHARP_IB_PORT``
See ``sharp_init_env()``
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

grom72 reviewed (commented):
> Reviewed 1 of 1 files at r11.
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: all files reviewed (commit messages unreviewed), all discussions resolved (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
> Reviewed 1 of 4 files at r5, 31 of 37 files at r6, 5 of 6 files at r7, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ldorau)



<!-- Sent from Reviewable.io -->

ldorau reviewed (commented):
@grom72 all requested changes have been done

> *[Reviewable](https://reviewable.io/reviews/grom72/libfabric/4#-:-NLLstOi-tw6PV8_H8qW:b-idjcoj)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @ldorau)



<!-- Sent from Reviewable.io -->
